File: typescript/packages/core/src/commands/cli/builtin/linear/reads.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import {
  getIssue,
  listIssueComments,
  listTeamCycles,
  listTeamDocuments,
  listTeamIssues,
  listTeamLabels,
  listTeamMembers,
  listTeamProjects,
  listTeams,
  resolveTeam,
  searchIssues,
  type LinearTransport,
} from '../../../../core/linear/_client.ts'
import type { LinearConfig } from '../../../../core/linear/config.ts'
import {
  buildProjectIssue,
  normalizeComment,
  normalizeCycle,
  normalizeDocument,
  normalizeIssue,
  normalizeLabel,
  normalizeProject,
  normalizeTeam,
  normalizeUser,
  toJsonBytes,
  type NormalizedProjectIssue,
} from '../../../../core/linear/normalize.ts'
import { FlagView } from '../../../spec/types.ts'
import { IOResult } from '../../../../io/types.ts'
import { enoent } from '../../../../utils/errors.ts'
import type { CommandFnResult } from '../../../config.ts'
import type { CLIInvocation } from '../../types.ts'
import { firstText, linearTransport, resolveIssue } from './util.ts'

function requireTeam(fl: FlagView): string {
  const team = fl.asStr('team')
  if (team === undefined || team === '') throw new Error('--team is required')
  return team
}

function teamStr(team: Record<string, unknown>, key: string): string | null {
  const value = team[key]
  return typeof value === 'string' ? value : null
}

async function teamProjectRows(
  transport: LinearTransport,
  teamId: string,
  projectId: string,
): Promise<NormalizedProjectIssue[]> {
  const teamIssues = await listTeamIssues(transport, teamId)
  const rows: NormalizedProjectIssue[] = []
  for (const issue of teamIssues) {
    const proj = issue.project
    const projObj =
      proj !== null && typeof proj === 'object' ? (proj as Record<string, unknown>) : {}
    if (projObj.id !== projectId) continue
    rows.push(buildProjectIssue(issue))
  }
  return rows
}

async function allUsers(transport: LinearTransport): Promise<Record<string, unknown>[]> {
  const teams = await listTeams(transport)
  const seen = new Set<string>()
  const users: Record<string, unknown>[] = []
  for (const team of teams) {
    const teamId = typeof team.id === 'string' ? team.id : ''
    for (const user of await listTeamMembers(transport, teamId)) {
      const uid = user.id
      if (typeof uid !== 'string' || seen.has(uid)) continue
      seen.add(uid)
      users.push(user)
    }
  }
  return users
}

export async function teamList(inv: CLIInvocation): Promise<CommandFnResult> {
  const cfg = inv.config as LinearConfig
  let teams = await listTeams(linearTransport(inv.config))
  if (cfg.teamIds !== undefined && cfg.teamIds.length > 0) {
    const keep = new Set(cfg.teamIds)
    teams = teams.filter((team) => typeof team.id === 'string' && keep.has(team.id))
  }
  return [toJsonBytes(teams.map((team) => normalizeTeam(team))), new IOResult()]
}

export async function teamGet(inv: CLIInvocation): Promise<CommandFnResult> {
  const team = await resolveTeam(linearTransport(inv.config), firstText(inv.texts, 'team key'))
  return [toJsonBytes(normalizeTeam(team)), new IOResult()]
}

export async function teamMembers(inv: CLIInvocation): Promise<CommandFnResult> {
  const transport = linearTransport(inv.config)
  const team = await resolveTeam(transport, firstText(inv.texts, 'team key'))
  const teamId = typeof team.id === 'string' ? team.id : ''
  const users = await listTeamMembers(transport, teamId)
  return [toJsonBytes(users.map((user) => normalizeUser(user))), new IOResult()]
}

export async function issueList(inv: CLIInvocation): Promise<CommandFnResult> {
  const transport = linearTransport(inv.config)
  const team = await resolveTeam(transport, requireTeam(new FlagView(inv.flags)))
  const teamId = typeof team.id === 'string' ? team.id : ''
  const issues = await listTeamIssues(transport, teamId)
  return [toJsonBytes(issues.map((issue) => normalizeIssue(issue))), new IOResult()]
}

export async function issueGet(inv: CLIInvocation): Promise<CommandFnResult> {
  const transport = linearTransport(inv.config)
  const issueId = await resolveIssue(transport, firstText(inv.texts, 'issue key'))
  const issue = await getIssue(transport, issueId)
  return [toJsonBytes(normalizeIssue(issue)), new IOResult()]
}

export async function projectList(inv: CLIInvocation): Promise<CommandFnResult> {
  const transport = linearTransport(inv.config)
  const team = await resolveTeam(transport, requireTeam(new FlagView(inv.flags)))
  const teamId = typeof team.id === 'string' ? team.id : ''
  const projects = await listTeamProjects(transport, teamId)
  const payload = []
  for (const project of projects) {
    const projectId = typeof project.id === 'string' ? project.id : ''
    const rows = await teamProjectRows(transport, teamId, projectId)
    payload.push(
      normalizeProject(project, {
        teamId,
        teamKey: teamStr(team, 'key'),
        teamName: teamStr(team, 'name'),
        issues: rows,
      }),
    )
  }
  return [toJsonBytes(payload), new IOResult()]
}

export async function projectGet(inv: CLIInvocation): Promise<CommandFnResult> {
  const transport = linearTransport(inv.config)
  const team = await resolveTeam(transport, requireTeam(new FlagView(inv.flags)))
  const teamId = typeof team.id === 'string' ? team.id : ''
  const projectId = firstText(inv.texts, 'project id')
  const projects = await listTeamProjects(transport, teamId)
  for (const project of projects) {
    if (project.id === projectId) {
      const rows = await teamProjectRows(transport, teamId, projectId)
      return [
        toJsonBytes(
          normalizeProject(project, {
            teamId,
            teamKey: teamStr(team, 'key'),
            teamName: teamStr(team, 'name'),
            issues: rows,
          }),
        ),
        new IOResult(),
      ]
    }
  }
  throw enoent(projectId)
}

export async function cycleList(inv: CLIInvocation): Promise<CommandFnResult> {
  const transport = linearTransport(inv.config)
  const team = await resolveTeam(transport, requireTeam(new FlagView(inv.flags)))
  const teamId = typeof team.id === 'string' ? team.id : ''
  const cycles = await listTeamCycles(transport, teamId)
  return [toJsonBytes(cycles.map((cycle) => normalizeCycle(cycle, teamId))), new IOResult()]
}

export async function cycleCurrent(inv: CLIInvocation): Promise<CommandFnResult> {
  const transport = linearTransport(inv.config)
  const team = await resolveTeam(transport, requireTeam(new FlagView(inv.flags)))
  const teamId = typeof team.id === 'string' ? team.id : ''
  const cycles = await listTeamCycles(transport, teamId)
  let current: Record<string, unknown> | undefined
  for (const cycle of cycles) {
    if (current === undefined) {
      current = cycle
      continue
    }
    const a = typeof cycle.number === 'number' ? cycle.number : 0
    const b = typeof current.number === 'number' ? current.number : 0
    if (a > b) current = cycle
  }
  if (current === undefined) throw enoent('cycles')
  return [toJsonBytes(normalizeCycle(current, teamId)), new IOResult()]
}

export async function cycleGet(inv: CLIInvocation): Promise<CommandFnResult> {
  const transport = linearTransport(inv.config)
  const team = await resolveTeam(transport, requireTeam(new FlagView(inv.flags)))
  const teamId = typeof team.id === 'string' ? team.id : ''
  const cycleId = firstText(inv.texts, 'cycle id')
  const cycles = await listTeamCycles(transport, teamId)
  for (const cycle of cycles) {
    if (cycle.id === cycleId) return [toJsonBytes(normalizeCycle(cycle, teamId)), new IOResult()]
  }
  throw enoent(cycleId)
}

export async function labelList(inv: CLIInvocation): Promise<CommandFnResult> {
  const transport = linearTransport(inv.config)
  const team = await resolveTeam(transport, requireTeam(new FlagView(inv.flags)))
  const teamId = typeof team.id === 'string' ? team.id : ''
  const labels = await listTeamLabels(transport, teamId)
  return [toJsonBytes(labels.map((label) => normalizeLabel(label))), new IOResult()]
}

export async function commentList(inv: CLIInvocation): Promise<CommandFnResult> {
  const transport = linearTransport(inv.config)
  const issueId = await resolveIssue(transport, firstText(inv.texts, 'issue key'))
  const issue = await getIssue(transport, issueId)
  const issueKey = typeof issue.identifier === 'string' ? issue.identifier : null
  const comments = await listIssueComments(transport, issueId)
  return [
    toJsonBytes(comments.map((comment) => normalizeComment(comment, issueId, issueKey))),
    new IOResult(),
  ]
}

export async function userList(inv: CLIInvocation): Promise<CommandFnResult> {
  const users = await allUsers(linearTransport(inv.config))
  return [toJsonBytes(users.map((user) => normalizeUser(user))), new IOResult()]
}

export async function userGet(inv: CLIInvocation): Promise<CommandFnResult> {
  const email = firstText(inv.texts, 'user email')
  for (const user of await allUsers(linearTransport(inv.config))) {
    if (user.email === email) return [toJsonBytes(normalizeUser(user)), new IOResult()]
  }
  throw enoent(email)
}

export async function documentList(inv: CLIInvocation): Promise<CommandFnResult> {
  const transport = linearTransport(inv.config)
  const team = await resolveTeam(transport, requireTeam(new FlagView(inv.flags)))
  const teamId = typeof team.id === 'string' ? team.id : ''
  const documents = await listTeamDocuments(transport, teamId)
  return [toJsonBytes(documents.map((document) => normalizeDocument(document))), new IOResult()]
}

export async function documentGet(inv: CLIInvocation): Promise<CommandFnResult> {
  const transport = linearTransport(inv.config)
  const team = await resolveTeam(transport, requireTeam(new FlagView(inv.flags)))
  const teamId = typeof team.id === 'string' ? team.id : ''
  const documentId = firstText(inv.texts, 'document id')
  const documents = await listTeamDocuments(transport, teamId)
  for (const document of documents) {
    if (document.id === documentId)
      return [toJsonBytes(normalizeDocument(document)), new IOResult()]
  }
  throw enoent(documentId)
}

export async function search(inv: CLIInvocation): Promise<CommandFnResult> {
  const fl = new FlagView(inv.flags)
  const flagQuery = fl.asStr('query')
  const query = flagQuery !== undefined && flagQuery !== '' ? flagQuery : (inv.texts[0] ?? null)
  if (query === null || query === '') throw new Error('a search query is required')
  const results = await searchIssues(linearTransport(inv.config), query)
  return [toJsonBytes(results), new IOResult()]
}

File: python/mirage/commands/cli/builtin/linear/reads.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import functools
from collections.abc import Awaitable, Callable

from mirage.commands.cli.builtin.linear.util import first_text, resolve_issue
from mirage.commands.cli.types import CLIInvocation
from mirage.commands.spec.types import FlagView
from mirage.core.linear._client import (get_issue, list_issue_comments,
                                        list_team_cycles, list_team_documents,
                                        list_team_issues, list_team_labels,
                                        list_team_members, list_team_projects,
                                        list_teams, resolve_team,
                                        search_issues)
from mirage.core.linear.config import LinearConfig
from mirage.core.linear.normalize import (normalize_comment, normalize_cycle,
                                          normalize_document, normalize_issue,
                                          normalize_label, normalize_project,
                                          normalize_team, normalize_user,
                                          to_json_bytes)
from mirage.io.stream import yield_bytes
from mirage.io.types import ByteSource, IOResult
from mirage.types import JsonValue


def _require_team(fl: FlagView) -> str:
    team = fl.as_str("team")
    if not team:
        raise ValueError("--team is required")
    return team


async def _project_issue_rows(config: LinearConfig, team_id: str,
                              project_id: str) -> list[dict[str, JsonValue]]:
    team_issues = await list_team_issues(config, team_id)
    rows: list[dict[str, JsonValue]] = []
    for issue in team_issues:
        if (issue.get("project") or {}).get("id") != project_id:
            continue
        state = issue.get("state") or {}
        rows.append({
            "issue_id": issue.get("id"),
            "issue_key": issue.get("identifier"),
            "title": issue.get("title"),
            "state_id": state.get("id"),
            "state_name": state.get("name"),
            "url": issue.get("url"),
        })
    return rows


async def _run_team_list(config: LinearConfig, texts: tuple[str, ...],
                         fl: FlagView) -> bytes:
    teams = await list_teams(config)
    if config.team_ids:
        teams = [t for t in teams if t.get("id") in config.team_ids]
    return to_json_bytes([normalize_team(team) for team in teams])


async def _run_team_get(config: LinearConfig, texts: tuple[str, ...],
                        fl: FlagView) -> bytes:
    team = await resolve_team(config, first_text(texts, "team key"))
    return to_json_bytes(normalize_team(team))


async def _run_team_members(config: LinearConfig, texts: tuple[str, ...],
                            fl: FlagView) -> bytes:
    team = await resolve_team(config, first_text(texts, "team key"))
    users = await list_team_members(config, team["id"])
    return to_json_bytes([normalize_user(user) for user in users])


async def _run_issue_list(config: LinearConfig, texts: tuple[str, ...],
                          fl: FlagView) -> bytes:
    team = await resolve_team(config, _require_team(fl))
    issues = await list_team_issues(config, team["id"])
    return to_json_bytes([normalize_issue(issue) for issue in issues])


async def _run_issue_get(config: LinearConfig, texts: tuple[str, ...],
                         fl: FlagView) -> bytes:
    issue_id = await resolve_issue(config, first_text(texts, "issue key"))
    issue = await get_issue(config, issue_id)
    return to_json_bytes(normalize_issue(issue))


async def _run_project_list(config: LinearConfig, texts: tuple[str, ...],
                            fl: FlagView) -> bytes:
    team = await resolve_team(config, _require_team(fl))
    projects = await list_team_projects(config, team["id"])
    payload = []
    for project in projects:
        rows = await _project_issue_rows(config, team["id"], project["id"])
        payload.append(
            normalize_project(project,
                              team_id=team["id"],
                              team_key=team.get("key"),
                              team_name=team.get("name"),
                              issues=rows))
    return to_json_bytes(payload)


async def _run_project_get(config: LinearConfig, texts: tuple[str, ...],
                           fl: FlagView) -> bytes:
    team = await resolve_team(config, _require_team(fl))
    project_id = first_text(texts, "project id")
    projects = await list_team_projects(config, team["id"])
    for project in projects:
        if project.get("id") == project_id:
            rows = await _project_issue_rows(config, team["id"], project_id)
            return to_json_bytes(
                normalize_project(project,
                                  team_id=team["id"],
                                  team_key=team.get("key"),
                                  team_name=team.get("name"),
                                  issues=rows))
    raise FileNotFoundError(project_id)


async def _run_cycle_list(config: LinearConfig, texts: tuple[str, ...],
                          fl: FlagView) -> bytes:
    team = await resolve_team(config, _require_team(fl))
    cycles = await list_team_cycles(config, team["id"])
    return to_json_bytes(
        [normalize_cycle(cycle, team_id=team["id"]) for cycle in cycles])


async def _run_cycle_current(config: LinearConfig, texts: tuple[str, ...],
                             fl: FlagView) -> bytes:
    team = await resolve_team(config, _require_team(fl))
    cycles = await list_team_cycles(config, team["id"])
    if not cycles:
        raise FileNotFoundError("no cycles")
    current = max(cycles, key=lambda cycle: cycle.get("number") or 0)
    return to_json_bytes(normalize_cycle(current, team_id=team["id"]))


async def _run_cycle_get(config: LinearConfig, texts: tuple[str, ...],
                         fl: FlagView) -> bytes:
    team = await resolve_team(config, _require_team(fl))
    cycle_id = first_text(texts, "cycle id")
    cycles = await list_team_cycles(config, team["id"])
    for cycle in cycles:
        if cycle.get("id") == cycle_id:
            return to_json_bytes(normalize_cycle(cycle, team_id=team["id"]))
    raise FileNotFoundError(cycle_id)


async def _run_label_list(config: LinearConfig, texts: tuple[str, ...],
                          fl: FlagView) -> bytes:
    team = await resolve_team(config, _require_team(fl))
    labels = await list_team_labels(config, team["id"])
    return to_json_bytes([normalize_label(label) for label in labels])


async def _run_comment_list(config: LinearConfig, texts: tuple[str, ...],
                            fl: FlagView) -> bytes:
    issue_id = await resolve_issue(config, first_text(texts, "issue key"))
    issue = await get_issue(config, issue_id)
    issue_key = issue.get("identifier")
    comments = await list_issue_comments(config, issue_id)
    return to_json_bytes([
        normalize_comment(comment, issue_id=issue_id, issue_key=issue_key)
        for comment in comments
    ])


async def _all_users(config: LinearConfig) -> list[dict[str, JsonValue]]:
    teams = await list_teams(config)
    seen: set[str] = set()
    users: list[dict[str, JsonValue]] = []
    for team in teams:
        for user in await list_team_members(config, team["id"]):
            uid = user.get("id")
            if not isinstance(uid, str) or uid in seen:
                continue
            seen.add(uid)
            users.append(user)
    return users


async def _run_user_list(config: LinearConfig, texts: tuple[str, ...],
                         fl: FlagView) -> bytes:
    users = await _all_users(config)
    return to_json_bytes([normalize_user(user) for user in users])


async def _run_user_get(config: LinearConfig, texts: tuple[str, ...],
                        fl: FlagView) -> bytes:
    email = first_text(texts, "user email")
    for user in await _all_users(config):
        if user.get("email") == email:
            return to_json_bytes(normalize_user(user))
    raise FileNotFoundError(email)


async def _run_document_list(config: LinearConfig, texts: tuple[str, ...],
                             fl: FlagView) -> bytes:
    team = await resolve_team(config, _require_team(fl))
    documents = await list_team_documents(config, team["id"])
    return to_json_bytes(
        [normalize_document(document) for document in documents])


async def _run_document_get(config: LinearConfig, texts: tuple[str, ...],
                            fl: FlagView) -> bytes:
    team = await resolve_team(config, _require_team(fl))
    document_id = first_text(texts, "document id")
    documents = await list_team_documents(config, team["id"])
    for document in documents:
        if document.get("id") == document_id:
            return to_json_bytes(normalize_document(document))
    raise FileNotFoundError(document_id)


async def _run_search(config: LinearConfig, texts: tuple[str, ...],
                      fl: FlagView) -> bytes:
    query = fl.as_str("query") or (texts[0] if texts else None)
    if not query:
        raise ValueError("a search query is required")
    results = await search_issues(config, query)
    return to_json_bytes(results)


Runner = Callable[[LinearConfig, tuple[str, ...], FlagView], Awaitable[bytes]]


async def _dispatch(
        runner: Runner, inv: CLIInvocation[LinearConfig]
) -> tuple[ByteSource | None, IOResult]:
    fl = FlagView(inv.flags)
    data = await runner(inv.config, inv.texts, fl)
    return yield_bytes(data), IOResult()


team_list = functools.partial(_dispatch, _run_team_list)
team_get = functools.partial(_dispatch, _run_team_get)
team_members = functools.partial(_dispatch, _run_team_members)
issue_list = functools.partial(_dispatch, _run_issue_list)
issue_get = functools.partial(_dispatch, _run_issue_get)
project_list = functools.partial(_dispatch, _run_project_list)
project_get = functools.partial(_dispatch, _run_project_get)
cycle_list = functools.partial(_dispatch, _run_cycle_list)
cycle_current = functools.partial(_dispatch, _run_cycle_current)
cycle_get = functools.partial(_dispatch, _run_cycle_get)
label_list = functools.partial(_dispatch, _run_label_list)
comment_list = functools.partial(_dispatch, _run_comment_list)
user_list = functools.partial(_dispatch, _run_user_list)
user_get = functools.partial(_dispatch, _run_user_get)
document_list = functools.partial(_dispatch, _run_document_list)
document_get = functools.partial(_dispatch, _run_document_get)
search = functools.partial(_dispatch, _run_search)
